File: Data Structures & Algorithms/valid-parenthesis-string/submission-3.cpp
class Solution {
private:
    vector<vector<int>> memo;

    bool dfs(string s, int sz, int idx) {
        if (idx >= s.size()) return sz == 0;

        if (memo[sz][idx] != 2) return memo[sz][idx] == 1;

        bool res = false;
        switch (s[idx]) {
            case '(':
                res |= dfs(s, sz + 1, idx + 1);
                break;
            case ')':
                if (sz <= 0) return memo[sz][idx] = 0;
                res |= dfs(s, sz - 1, idx + 1);
                break;
            case '*':
                res |= dfs(s, sz + 1, idx + 1);
                
                if (sz > 0) res |= dfs(s, sz - 1, idx + 1);
                
                res |= dfs(s, sz, idx + 1);
        }

        return memo[sz][idx] = (int)res;
    }

public:
    bool checkValidString(string s) {
        memo.assign(s.size(), vector<int>(s.size(), 2)); // 0 = false, 1 = true, 2 = idk
        return dfs(s, 0, 0);
    }
};

File: Data Structures & Algorithms/interleaving-string/submission-3.py
from functools import cache


class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        if len(s3) != len(s1) + len(s2):
            return False
            
        @cache
        def dfs(i, j):
            k = i + j
            if k >= len(s3):
                return True
            if i < len(s1) and s3[k] == s1[i]:
                if dfs(i + 1, j):
                    return True
            if j < len(s2) and s3[k] == s2[j]:
                if dfs(i, j + 1):
                    return True
            return False

        return dfs(0, 0)
